Update implementation of contain flag for motion path <ray>

nmoucht opened the pull request:
#### 7e206701112298ed0df6f7a8e3a393e9e8531530
<pre>
Update implementation of contain flag for motion path &lt;ray&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=256225">https://bugs.webkit.org/show_bug.cgi?id=256225</a>
rdar://108861663

Reviewed by Tim Nguyen.

After w3c/fxtf-drafts@bd9fb9b, we need to update our contain
implementation to follow this new behavior. This involves replacing the old implementation
in RayPathOperation::lengthForContainPath to the new behavior, which just bases the length
we shorten the computed path length on half the width or height of the element that we are
animating. Since we no longer need the anchor point or the angle of the ray, remove this
from RayPathOperation::lengthForContainPath&apos;s function parameters (as well as all other
PathOperations).

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-contain-004.html:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
(WebCore::AcceleratedEffectValues::AcceleratedEffectValues):
* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::RayPathOperation::lengthForContainPath const):
(WebCore::RayPathOperation::getPath const):
* Source/WebCore/rendering/PathOperation.h:
(WebCore::PathOperation::getPath):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::applyMotionPathTransform const):

Canonical link: <a href="https://commits.webkit.org/265106@main">https://commits.webkit.org/265106@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f354e141c00f5de778c7587efbe08797d20d6dfb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9824 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10072 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10317 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11476 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9552 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9833 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12056 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10022 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12482 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9978 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10785 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8330 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/11886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8092 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8910 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/11886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9190 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9059 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/11886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9559 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7784 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8725 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8807 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2357 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12951 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9335 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/PathOperation.cpp:
```diff
@@ -121,62 +121,19 @@ double RayPathOperation::lengthForPath() const
-    
-    // TODO: Implement minimally increasing path length to allow all vertices to be within such a path length
-    return computedPathLength;
+    return computedPathLength - (std::min(elementRect.width(), elementRect.height()) / 2);
```
The spec says:
> the path’s length is reduced by half the width or half the height of the element’s border box, whichever is larger, and floored at zero.

So should this be:

```
return std::max(0.0, computedPathLength - (std::max(elementRect.width(), elementRect.height()) / 2));
```

nt1m replied:
I can confirm this fixes the remaining WPT

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/PathOperation.cpp:
```diff
@@ -121,62 +121,19 @@ double RayPathOperation::lengthForPath() const
 }
 
-const std::optional<Path> RayPathOperation::getPath(const FloatRect& referenceRect, FloatPoint anchor, OffsetRotation rotation) const
+const std::optional<Path> RayPathOperation::getPath(const FloatRect& referenceRect, FloatPoint, OffsetRotation) const
```
We could probably get rid of the two unused parameters altogether, all the other `PathOperation` classes don't use them.

nmoucht replied:
Fixed.

nt1m reviewed (approved):
The changelog needs updating, but looks good otherwise, thanks!

nmoucht committed:
Update implementation of contain flag for motion path <ray>
https://bugs.webkit.org/show_bug.cgi?id=256225
rdar://108861663

Reviewed by Tim Nguyen.

After w3c/fxtf-drafts@bd9fb9b, we need to update our contain
implementation to follow this new behavior. This involves replacing the old implementation
in RayPathOperation::lengthForContainPath to the new behavior, which just bases the length
we shorten the computed path length on half the width or height of the element that we are
animating. Since we no longer need the anchor point or the angle of the ray, remove this
from RayPathOperation::lengthForContainPath's function parameters (as well as all other
PathOperations).

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-contain-004.html:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
(WebCore::AcceleratedEffectValues::AcceleratedEffectValues):
* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::RayPathOperation::lengthForContainPath const):
(WebCore::RayPathOperation::getPath const):
* Source/WebCore/rendering/PathOperation.h:
(WebCore::PathOperation::getPath):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::applyMotionPathTransform const):

Canonical link: https://commits.webkit.org/265106@main
